plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/service/QueryGroupPersistenceService.java

```diff
@@ -0,0 +1,183 @@
+        }
+        this.maxQueryGroupCount = newMaxQueryGroupCount;
+    }
+
+    /**
+     * Update cluster state to include the new QueryGroup
+     * @param queryGroup {@link QueryGroup} - the QueryGroup we're currently creating
+     * @param listener - ActionListener for CreateQueryGroupResponse
+     */
+    public void persistInClusterStateMetadata(QueryGroup queryGroup, ActionListener<CreateQueryGroupResponse> listener) {
+        clusterService.submitStateUpdateTask(SOURCE, new ClusterStateUpdateTask(Priority.NORMAL) {
+            @Override
+            public ClusterState execute(ClusterState currentState) throws Exception {
+                return saveQueryGroupInClusterState(queryGroup, currentState);
+            }
+
+            @Override
+            public ThrottlingKey getClusterManagerThrottlingKey() {
+                return createQueryGroupThrottlingKey;
+            }
+
+            @Override
+            public void onFailure(String source, Exception e) {
+                logger.warn("failed to save QueryGroup object due to error: {}, for source: {}", e.getMessage(), source);
+                listener.onFailure(e);
+            }
+
+            @Override
+            public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+                CreateQueryGroupResponse response = new CreateQueryGroupResponse(queryGroup, RestStatus.OK);
+                listener.onResponse(response);
+            }
+        });
+    }
+
+    /**
+     * This method will be executed before we submit the new cluster state
+     * @param queryGroup - the QueryGroup we're currently creating
+     * @param currentClusterState - the cluster state before the update
+     */
+    ClusterState saveQueryGroupInClusterState(final QueryGroup queryGroup, final ClusterState currentClusterState) {
+        final Metadata metadata = currentClusterState.metadata();
+        String groupName = queryGroup.getName();
+        final Map<String, QueryGroup> previousGroups = metadata.queryGroups();
+
+        // check if maxQueryGroupCount will breach
+        if (previousGroups.size() + 1 > maxQueryGroupCount) {
+            logger.error("{} value exceeded its assigned limit of {}", QUERY_GROUP_COUNT_SETTING_NAME, maxQueryGroupCount);
+            throw new RuntimeException("Can't create more than " + maxQueryGroupCount + " QueryGroups in the system");
+        }
+
+        // check for duplicate name
+        Optional<QueryGroup> findExistingGroup = previousGroups.values()
+            .stream()
+            .filter(group -> group.getName().equals(groupName))
+            .findFirst();
+        if (findExistingGroup.isPresent()) {
+            logger.warn("QueryGroup with name {} already exists. Not creating a new one.", groupName);
+            throw new RuntimeException("QueryGroup with name " + groupName + " already exists. Not creating a new one.");
+        }
```

Comment:
Any validations on the characters/length allowed in QueryGroup name? That validation can be done preemptively while accepting the request

Reply:
We have this in the querygroup constructor 
```
if (name.length() > MAX_CHARS_ALLOWED_IN_NAME) {
            throw new IllegalArgumentException("QueryGroup.name shouldn't be more than 50 chars long");
        }
```